### DISCUSSION TITLE
Query related to code development in peridigm

### DISCUSSION THREAD

**Discussion by deepakbiitk** [peridigm/peridigm#343] [Q&A]
Hi all, 

I wrote contact model in the contact folder and registered it in Peridigm_ContactModelFactory.cpp. I complied the code using make -j 4. I received the following error.

../../libPeridigmLib.a(Peridigm_ContactModelFactory.cpp.o): In function `PeridigmNS::ContactModelFactory::create(Teuchos::ParameterList const&)':
Peridigm_ContactModelFactory.cpp:(.text+0x209): undefined reference to `PeridigmNS::RepulsiveForceContactModel::RepulsiveForceContactModel(Teuchos::ParameterList const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [src/compute/unit_test/utPeridigm_Compute_Angular_Momentum] Error 1
make[1]: *** [src/compute/unit_test/CMakeFiles/utPeridigm_Compute_Angular_Momentum.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
../../libPeridigmLib.a(Peridigm_ContactModelFactory.cpp.o): In function `PeridigmNS::ContactModelFactory::create(Teuchos::ParameterList const&)':
Peridigm_ContactModelFactory.cpp:(.text+0x209): undefined reference to `PeridigmNS::RepulsiveForceContactModel::RepulsiveForceContactModel(Teuchos::ParameterList const&)'
collect2: error: ld returned 1 exit status
libPeridigmLib.a(Peridigm_ContactModelFactory.cpp.o): In function `PeridigmNS::ContactModelFactory::create(Teuchos::ParameterList const&)':
Peridigm_ContactModelFactory.cpp:(.text+0x209): undefined reference to `PeridigmNS::RepulsiveForceContactModel::RepulsiveForceContactModel(Teuchos::ParameterList const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [src/compute/unit_test/utPeridigm_Compute_Force] Error 1
make[1]: *** [src/compute/unit_test/CMakeFiles/utPeridigm_Compute_Force.dir/all] Error 2
make[2]: *** [src/Peridigm] Error 1
make[1]: *** [src/CMakeFiles/Peridigm.dir/all] Error 2
../../libPeridigmLib.a(Peridigm_ContactModelFactory.cpp.o): In function `PeridigmNS::ContactModelFactory::create(Teuchos::ParameterList const&)':
Peridigm_ContactModelFactory.cpp:(.text+0x209): undefined reference to `PeridigmNS::RepulsiveForceContactModel::RepulsiveForceContactModel(Teuchos::ParameterList const&)'
collect2: error: ld returned 1 exit status
make[2]: *** [src/compute/unit_test/utPeridigm_Compute_Linear_Momentum] Error 1
make[1]: *** [src/compute/unit_test/CMakeFiles/utPeridigm_Compute_Linear_Momentum.dir/all] Error 2
make: *** [all] Error 2

i am trying to debug the error. I appreciate any help regarding this.

Thanks,
Deepak Behera

**Comment by jritta** [comment]
Hi Deepak,

Did you include the header file of your new contact model in `Peridigm_ContactModelFactory.cpp`?

Best
Jonas

**Comment by deepakbiitk** [comment]
Hi Jonas,

Yes, I did include the header file in Peridigm_ContactModelFactory.cpp

#include <Teuchos_Assert.hpp>
#include "Peridigm_ContactModelFactory.hpp"
#include "Peridigm_ShortRangeForceContactModel.hpp"
#include "Peridigm_RepulsiveForceContactModel.hpp"
#include "Peridigm_UserDefinedTimeDependentShortRangeForceContactModel.hpp"

using namespace std;

Teuchos::RCP<PeridigmNS::ContactModel>
PeridigmNS::ContactModelFactory::create(const Teuchos::ParameterList& contactModelParams)
{
  string contactModelName = contactModelParams.get<string>("Contact Model");

  Teuchos::RCP<PeridigmNS::ContactModel> contactModel;
  if (contactModelName == "Short Range Force")
    contactModel = Teuchos::rcp( new ShortRangeForceContactModel(contactModelParams) );
  else if (contactModelName == "Repulsive Force")
    contactModel = Teuchos::rcp( new RepulsiveForceContactModel(contactModelParams) ); 
  else if (contactModelName == "Time Dependent Short Range Force")
    contactModel = Teuchos::rcp( new UserDefinedTimeDependentShortRangeForceContactModel(contactModelParams) );  
  else {
    string invalidContactModel("\n**** Unrecognized contact model: ");
    invalidContactModel += contactModelName;
    invalidContactModel += ", must be \"Short Range Force\" or \"Time Dependent Short Range Force\".\n";
    TEUCHOS_TEST_FOR_EXCEPT_MSG(true, invalidContactModel);
  }
  
  return contactModel;
}

Do you know are there any other files, where I should include my header file?

Best Regards,
Deepak

**Comment by jritta** [comment]
Hi Deepak,

It could be that you also have to include it in `Peridigm.cpp`, but I am not 100% sure about that because I never implemented a contact model so far.

Is the error message you posted the complete error message or is there already an error beforehand?

In the case of a compile error I also suggest to compile with only one thread such that the compiler output is easier to read and analyse.

Best
Jonas

**Reply by deepakbiitk** [reply]
I included it in Peridigm.cpp. I did get the same error.  

Thanks for the help Jonas.

Best,
Deepak

**Reply by jritta** [reply]
Ok, unfortunately I can not give further advice without additional information.

Maybe you can share your code (if you don't want to fully publish maybe with mock functions) such that we can help you.

Best,
Jonas
